src/Microsoft.AspNetCore.Authentication/Events/BaseContext.cs

```diff
@@ -9,41 +9,70 @@ namespace Microsoft.AspNetCore.Authentication
     /// <summary>
     /// Base class used by other context classes.
     /// </summary>
-    public abstract class BaseContext
+    public abstract class BaseContext<TOptions> where TOptions : AuthenticationSchemeOptions
     {
+        private AuthenticationProperties _properties;
+
         /// <summary>
         /// Constructor.
         /// </summary>
-        /// <param name="context">The request context.</param>
-        protected BaseContext(HttpContext context)
+        /// <param name="context">The context.</param>
+        /// <param name="scheme">The authentication scheme.</param>
+        /// <param name="options">The authentication options associated with the scheme.</param>
+        protected BaseContext(HttpContext context, AuthenticationScheme scheme, TOptions options)
         {
             if (context == null)
             {
                 throw new ArgumentNullException(nameof(context));
             }
+            if (scheme == null)
+            {
+                throw new ArgumentNullException(nameof(scheme));
+            }
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
 
             HttpContext = context;
+            Scheme = scheme;
+            Options = options;
         }
 
         /// <summary>
+        /// The authentication scheme.
+        /// </summary>
+        public AuthenticationScheme Scheme { get; }
+
+        /// <summary>
+        /// Gets the authentication options associated with the scheme.
+        /// </summary>
+        public TOptions Options { get; }
+
+        /// <summary>
         /// The context.
         /// </summary>
         public HttpContext HttpContext { get; }
 
         /// <summary>
         /// The request.
         /// </summary>
-        public HttpRequest Request
-        {
-            get { return HttpContext.Request; }
-        }
+        public HttpRequest Request => HttpContext.Request;
 
         /// <summary>
         /// The response.
         /// </summary>
-        public HttpResponse Response
+        public HttpResponse Response => HttpContext.Response;
+
+        /// <summary>
+        /// Gets or sets the <see cref="AuthenticationProperties"/>.
+        /// </summary>
+        public virtual AuthenticationProperties Properties
         {
-            get { return HttpContext.Response; }
+            get => _properties ?? (_properties = new AuthenticationProperties());
+            set => _properties = value;
         }
+
+        public EventResultState State { get; protected set; }
```

Comment:
Yeah so I'm not saying go back to only a ticket, I'm saying HandleRequestContext, has a RemoteAuthenticationResult, you can keep the method split like you have it where HandleRequestContext has Handle and Skip, and RemoteAuthenticateResultContext (maybe just shorten this to RemoteResultContext too) has Complete, Noop, Reject)

```
class HandleRequestContext {
    RemoteAuthenticationResult Result { get; set; };
    void Handle() =>Result = RemoteAuthenticationResult.Handled();
    void Skip() => Result = RemoteAuthenticationResult.Skip();
} 

class RemoteResultContext : HandleRequestContext {
   Complete/Success(AuthenticationTicket ticket) => Result = RemoteAuthenticationTicket.Success(ticket)
   Reject/Fail(failure) => Result = RemoteAuthenticationTicket.Fail(failure);
   None() => Result = RemoteAuthenticationTicket.None();

   // bool IsProcessingComplete() is replaced by checking if context.Result != null
}
```

We almost don't need the methods at all, and could just let them directly set the Result if we didn't want to hide some methods from some contexts